### PR TITLE
feat: XYNE-64084 Scheduled Call Channel Pill

### PR DESCRIPTION
**Pull request by ishaanrawat-netizen**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Areas Touched

- [ ] `apps/backend` (API / worker)
- [ ] `apps/dashboard`
- [ ] `apps/electron`
- [ ] `apps/xyne-claw` / `apps/xyne-claw-auth`
- [ ] `packages/shared` or another shared package
- [ ] Infra (docker, scripts, nix, CI)

---

## Zero (Sync) Changes

- [ ] No Zero changes — **skip this section**
- [ ] Adds/modifies a **mutator**
- [ ] Adds/modifies the **schema** (tables, columns, relationships)
- [ ] Adds/modifies a **query or ACL**

If any of the above:

- [ ] Server (`apps/backend/src/zero/mutators.ts`) and client (`apps/dashboard/src/zero/mutators.ts`) mutators mirror each other — same namespace, name, and args schema
- [ ] No `Date.now()` / `uuid()` generated inside a mutator
- [ ] Optimistic client result matches the server result (no state flash on rebase)
- [ ] New tables exported in `createSchema()` and given a Query ACL registered in `query-acl-factory.ts`
- [ ] Matching Prisma model + migration, client regenerated

## Schema & Data Changes

- [ ] No schema changes — **skip this section**
- [ ] Prisma schema modified (`apps/backend/prisma` and/or `prisma-common`)
- [ ] Migration added
- [ ] Backfill / data migration included

If any of the above:

- [ ] Clients regenerated (`db:generate`, `db:common:generate`)
- [ ] No new Postgres enum or enum value — enums are frozen (`pnpm run test:enum`)
- [ ] Backward compatible with deployed code, or rollout order noted below
- [ ] Backfill is idempotent

## Config, Environment & URLs

- [ ] No config changes — **skip this section**
- [ ] Env variable added / renamed / removed
- [ ] **Base URL, host, port, or origin changed** (API, Zero, Vespa, Claw, LiveKit, LiteLLM, …)
- [ ] CORS / allowed origins / OAuth redirect URIs changed
- [ ] Feature flag or runtime config changed

If any of the above:

- [ ] Key added to **all** relevant env files, not just the one you run — `apps/backend/.env.{example,local,test}`, `apps/dashboard/.env.{example,local,test}`, plus the app you touched
- [ ] Env setup / secret scripts and docker compose updated
- [ ] Verified in every consumer with its own base URL — dashboard (`VITE_API_BASE_URL`, `VITE_ZERO_SERVER`), Electron (`VITE_ELECTRON_*`), shareable links (`VITE_SHAREABLE_ORIGIN`), server-to-server (`BACKEND_URL`, `XYNE_CLAW_URL`, `XYNE_CLAW_AUTH_URL`)
- [ ] Google / Microsoft OAuth redirect URIs still resolve
- [ ] Nothing hardcoded, no secrets committed

## API Contract

- [ ] No API changes
- [ ] New endpoint
- [ ] Existing contract modified (shape, status codes, auth)
- [ ] Breaking for dashboard / electron / external / bots — migration noted above

---

### Automated

- [ ] Backend unit tests — `pnpm --filter xyne-spaces-backend run test`
- [ ] Claw tests — `pnpm --filter xyne-claw run test`
- [ ] End-to-end suite — `pnpm run test`
- [ ] Added / updated tests for this change
- [ ] Not covered by tests 

### Manual

**Users**
- [ ] Single user
- [ ] Two users at once (two accounts / two browsers) — sync lands on both, no cross-user leakage
- [ ] Different roles or permission levels (member vs admin, ACL boundaries)
- [ ] Different workspaces / spaces — data stays scoped

**Login**
- [ ] Google
- [ ] Microsoft
- [ ] Dev auth (`ENABLE_DEV_AUTH`)
- [ ] Fresh signup / first-time user
- [ ] Session expiry, re-login, logout

**Run mode**
- [ ] Local dev (`pnpm run dev:all`)
- [ ] Test mode (`dev:test`)
- [ ] Sandbox / claw test (`claw:test`)
- [ ] Prod-like local (`dev:prod`)
- [ ] Docker compose
- [ ] Electron desktop

**Sync & resilience** — for anything touching Zero
- [ ] Multiple tabs stay consistent
- [ ] Reload / hard refresh — no duplicate or lost rows
- [ ] Offline then reconnect — pending mutations replay cleanly
- [ ] Concurrent edits on the same entity by two users

**Surface & data**
- [ ] Web browser
- [ ] Electron desktop
- [ ] Narrow viewport, dark + light theme
- [ ] Empty state
- [ ] Large / realistic data volume
- [ ] Error paths (network failure, denied permission, invalid input)

---

## Checklist

- [ ] Reviewed my own diff; scoped to one thing
- [ ] `pnpm run build:shared` passes
- [ ] Backend `typecheck` + `build` pass
- [ ] Dashboard `lint:errors-only` + `typecheck` + `build` pass
- [ ] Formatting clean in the packages I touched
- [ ] New deps added with `pnpm --filter <pkg> add <dep>`; version pins in root `pnpm.overrides`
- [ ] Commits follow `<type>: <TICKET-ID> <subject>`; branch is `fix/*` or `feature/*`
- [ ] Docs / guidelines updated where behaviour changed
- [ ] No debug logging or commented-out code left behind

### CODE REVIEW

**Comment by ishaanrawat-netizen** [comment]
Screenshot 2026-09-17 at 13 01 20

File: apps/backend/src/controllers/scheduleCallController.ts [hunk 1]
```diff
@@ -34,6 +35,88 @@ function defaultCallInvitation(timezone?: string | null): CallInvitationParams['
...
+    // A thread-scheduled call's pill lives inside its thread, which does not travel with
+    // the call's channel. Leave it where it is. Keyed on the pill's own placement flag,
+    // not on metadata.conversationId — activateScheduledCall stamps that too, so a call
+    // that went live and bounced back to SCHEDULED would otherwise look thread-scheduled.
+    if (metadata?.channelPillInThread) return;
+
+    const workspaceId = await repositories.channels.getWorkspaceId(newChannelId);
+    const organizer = await repositories.users.findById(call.createdByUserId);
+    const organizerName = organizer?.displayName || organizer?.name || 'Someone';
+
```
Comment (XyneSpaces) [inline apps/backend/src/controllers/scheduleCallController.ts:96]: **[🟠 High] #105** — `metadata.channelPillMessageId` is read before the transaction in `moveScheduledCallPill`; a concurrent channel edit can retire a stale pill id and leave an orphaned live pill in the source channel. Re-read the pill message id (or the whole metadata row) inside the transaction, or select-for-update the call row, before retiring.

File: apps/backend/src/controllers/scheduleCallController.ts [hunk 2]
```diff
@@ -34,6 +35,88 @@ function defaultCallInvitation(timezone?: string | null): CallInvitationParams['
...
+   * Whether a call gets a read-only "upcoming call" pill in its channel.
+   *
+   * Recurring series are excluded: instances are materialized 60 days ahead and
+   * replenished forever, so a series would drip a fresh card into the channel for
+   * every occurrence. External calendar events are excluded because they are not ours
+   * to surface — and the test has to be on `callOrigin`, not on whether a channelId is
+   * set, because `xyneManaged` events carry a real one (googleCalendarCallStore:156).
+   */
+  private shouldPostScheduledCallPill(call: {
+    callOrigin: string;
```
Comment (XyneSpaces) [inline apps/backend/src/controllers/scheduleCallController.ts:48]: **[🟠 High] #204** — New scheduled-call pill lifecycle (`shouldPostScheduledCallPill`, `createScheduledCallPill`, `moveScheduledCallPill`, `retireScheduledCallPill`) has no tests. Add unit/integration tests covering gating (recurring/external/headless excluded), move race scenarios, pill creation, and retirement metadata.

File: apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.tsx [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+  message: {
+    messageId: string;
+    metadata: MessageMetadata | null;
+  };
+  /** The call's public externalId, from the message metadata. */
+  callId: string;
+  channelId: string;
+}
+
+export function ScheduledCallPill({
```
Comment (XyneSpaces) [inline apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.tsx:136]: **[🟠 High] #204** — The new `ScheduledCallPill` component and helpers (`toPillState`, `formatWhen`, join/reschedule/summary state mapping) have no tests. Add render tests per state and unit tests for the state/format helpers.

File: apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.tsx [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+              <span className={cn(state === 'CANCELLED' && 'xs-cc__meta--struck')}>{when}</span>
+            )}
+          </span>
+        </div>
+
+        <button
+          type='button'
+          onClick={isReschedule ? openInCalls : isOpenSummary ? openSummary : handleJoin}
+          disabled={!isJoinable && !isReschedule && !isOpenSummary}
+          className={cn('xs-cc__btn', isActive ? 'xs-cc__btn--call' : 'xs-cc__btn--secondary')}
```
Comment (XyneSpaces) [inline apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.tsx:287]: **[🟢 Low] #285** — `disabled={!isJoinable && !isReschedule && !isOpenSummary}` is always `false` because every `PillState` sets exactly one of these flags; the prop has no effect. Remove the dead condition or use a meaningful guard.

File: apps/dashboard/src/global.css [hunk 5]
```diff
@@ -5241,3 +5241,188 @@ img.inline-emoji,
...
+.xs-cc-scope {
+  --cc-bg: #ffffff;
+  --cc-bg2: #f7f8f9;
+  --cc-bd: #e4e7ea;
+  --cc-t1: #1f2329;
+  --cc-t2: #3a4048;
+  --cc-t4: #858c95;
+  --cc-t5: #9ca3ab;
+  --cc-t6: #b3b9c0;
+  /* Call green — the design system's --call family, not --status-success. */
```
Comment (XyneSpaces) [inline apps/dashboard/src/global.css:5262]: **[🟡 Medium] #182** — Hardcoded hex palette only overrides `[data-theme="midnight"]"; the app also supports `summer_breeze`, so that theme falls back to the light classic ramp. Either extend the override block to cover `summer_breeze` or map the pill tokens to existing theme variables for all themes.

**Review by XyneSpaces — COMMENTED** [review]
See the top-level summary comment for the verdict, reviewer recommendation, and the full findings table. Inline comments are posted below.

**Comment by XyneSpaces** [comment]
## Xyne Spaces Reviewer — PR #1972

**Verdict:** Changes requested after addressing the 🟠 High findings
**Reviewed:** 9 files, 14 hunks · **Findings:** 0 blocking, 2 high, 2 medium, 1 nit

### Findings
| Severity | Rule | Location | Issue |
|----------|------|----------|-------|
| 🟠 High | #105 | apps/backend/src/controllers/scheduleCallController.ts:~96 | Race in `moveScheduledCallPill`: metadata read outside transaction can retire stale pill id |
| 🟠 High | #204 | apps/backend/src/controllers/scheduleCallController.ts:~48 | No tests for new backend pill lifecycle |
| 🟠 High | #204 | apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.tsx:136 | No tests for `ScheduledCallPill` component & helpers |
| 🟡 Medium | #155 | apps/dashboard/src/global.css:~5410 | Button hover styles lack `:focus-visible` equivalents |
| 🟡 Medium | #182 | apps/dashboard/src/global.css:~5262 | Hardcoded palette misses `summer_breeze` theme |
| 🟢 Low | #285 | apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.tsx:287 | `disabled` prop is always false (dead logic) |

### Recommended reviewer
**@abhinavjuspay** (mentioned; reviewer-request API is not available to this bot — please assign manually) — owns the call-UI surface adjacent to `MessageBubble`/`ChatBubble`, with the lightest current review load.

| Candidate | Owns (files in this PR) | Default reviewer? | Pending GitHub reviews | Tie-breaker |
|-----------|-------------------------|-------------------|------------------------|-------------|
| @abhinavjuspay | `ScheduledCallPill.tsx`, `MessageBubble.tsx`, `ChatBubble.tsx` | no (file owner) | 0 | lightest Spaces workload (9 open tickets); recent call-UI commit |
| Mayank Bansal | `scheduleCallController.ts`, `callRepository.ts` | no (de facto backend owner) | unknown (GitHub login not discoverable) | strongest backend owner per Spaces context |
| Shubham Dhamija | hybrid call scheduling + UI | no | — | 82 open tickets |

Top issue: the concurrency race in `moveScheduledCallPill` — re-read metadata inside the transaction before retiring.

Posted 6 inline comments.

File: apps/dashboard/src/global.css [hunk 6]
```diff
@@ -5241,3 +5241,188 @@ img.inline-emoji,
...
+  letter-spacing: -0.1px;
+  cursor: pointer;
+  white-space: nowrap;
+  transition:
+    background 0.12s,
+    border-color 0.12s,
+    opacity 0.12s;
+}
+.xs-cc__btn:disabled {
+  opacity: 0.5;
```
Comment (XyneSpaces) [inline apps/dashboard/src/global.css:5410]: **[🟡 Medium] #155** — Button hover styles (`.xs-cc__btn--secondary:hover:not(:disabled)`, `.xs-cc__btn--call:hover:not(:disabled)`) lack matching `:focus-visible` states, so keyboard users only get the default browser outline. Add explicit `:focus-visible` rules with the same visual treatment.

**Comment by XyneSpaces** [comment]
**Reviewer recommendation update** (after re-checking Spaces context)

- **Lead backend reviewer:** Mayank Bansal (`mayank.bansal@juspay.in`) — strongest `scheduleCallController` / call-core context; explicitly asked for review of a PR touching `scheduleCallController.ts` in `#xyne-spaces-pr-review` [clf-functions.Xyne_Spaces__spaces-search-v2:3#1]. His GitHub handle is not discoverable from Spaces, so please assign him manually if he has repo access.
- **Resolvable UI/chat reviewer:** Satvik Batra (`@satvikbatra`, `satvik.batra@juspay.in`) — has worked on `MessageBubble.tsx`/`ChatBubble.tsx` and reported the ACL bug in `scheduleCallController`/`callController` [clf-functions.Xyne_Spaces__spaces-search-v2:3#6] [clf-functions.Xyne_Spaces__spaces-search-v2:3#7]; also asked to review GitHub UI PRs.

My earlier fallback to @abhinavjuspay is superseded by this mapping. Please prioritize Satvik for the GitHub review request and route the backend/core scheduling logic to Mayank.

File: apps/backend/src/controllers/scheduleCallController.ts [hunk 7]
```diff
@@ -34,6 +35,88 @@ function defaultCallInvitation(timezone?: string | null): CallInvitationParams['
...
+  /**
+   * Whether a call gets a read-only "upcoming call" pill in its channel.
+   *
+   * Recurring series are excluded: instances are materialized 60 days ahead and
+   * replenished forever, so a series would drip a fresh card into the channel for
+   * every occurrence. External calendar events are excluded because they are not ours
+   * to surface — and the test has to be on `callOrigin`, not on whether a channelId is
+   * set, because `xyneManaged` events carry a real one (googleCalendarCallStore:156).
+   */
+  private shouldPostScheduledCallPill(call: {
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/controllers/scheduleCallController.ts]: move these to helper and db operations to respective repositories

File: apps/backend/src/controllers/scheduleCallController.ts [hunk 8]
```diff
@@ -34,6 +35,88 @@ function defaultCallInvitation(timezone?: string | null): CallInvitationParams['
...
+      title: string | null;
+      createdByUserId: string;
+    };
+    newChannelId: string;
+  }): Promise<void> {
+    const { call, newChannelId } = params;
+    if (!this.shouldPostScheduledCallPill(call)) return;
+
+    const metadata = call.metadata as
+      | { channelPillMessageId?: string; channelPillInThread?: boolean }
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/controllers/scheduleCallController.ts]: why this channelPillInThread is needed?

File: apps/backend/src/controllers/scheduleCallController.ts [hunk 9]
```diff
@@ -34,6 +35,88 @@ function defaultCallInvitation(timezone?: string | null): CallInvitationParams['
...
+    // the call's channel. Leave it where it is. Keyed on the pill's own placement flag,
+    // not on metadata.conversationId — activateScheduledCall stamps that too, so a call
+    // that went live and bounced back to SCHEDULED would otherwise look thread-scheduled.
+    if (metadata?.channelPillInThread) return;
+
+    const workspaceId = await repositories.channels.getWorkspaceId(newChannelId);
+    const organizer = await repositories.users.findById(call.createdByUserId);
+    const organizerName = organizer?.displayName || organizer?.name || 'Someone';
+
+    const { conversationId } = await DatabaseClient.getInstance().$transaction(async (tx) => {
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/controllers/scheduleCallController.ts]: move these to repository

File: apps/backend/src/controllers/scheduleCallController.ts [hunk 10]
```diff
@@ -34,6 +35,88 @@ function defaultCallInvitation(timezone?: string | null): CallInvitationParams['
...
+   */
+  private shouldPostScheduledCallPill(call: {
+    callOrigin: string;
+    callType: string;
+    isRecurring: boolean;
+    recurringSeriesId: string | null;
+  }): boolean {
+    if (call.isRecurring || call.recurringSeriesId) return false;
+    if (call.callType === CallType.HEADLESS) return false;
+    return call.callOrigin === CallOrigin.CHANNEL || call.callOrigin === CallOrigin.CONVERSATION;
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/controllers/scheduleCallController.ts]: these are the only origin here so this check is not needed itself ig?

File: apps/backend/src/database/repositories/callRepository.ts [hunk 11]
```diff
@@ -63,6 +73,14 @@ export interface CallMetadata {
   conversationId?: string;
   artifactMessageId?: string;
   googleCalendarPush?: GoogleCalendarPushState;
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/database/repositories/callRepository.ts:98]: please remove these big comments, from everywhere

File: apps/backend/src/zero/queries.ts [hunk 12]
```diff
@@ -2868,6 +2869,38 @@ export const queries: AnyQueryRegistry = defineQueries({
...
+   *
+   * Must stay in sync with the copy in packages/shared/src/zero/queries.ts — zero-cache
+   * resolves named queries against THIS registry (ZERO_QUERY_URL → /api/zero/query), so
+   * a query defined only in shared silently returns whatever the client already happens
+   * to hold locally, which differs per user.
+   *
+   * The pill message is only an anchor: everything the card renders — title, organizer,
+   * times, status — is read from here, which is what lets a cancel or a title/time edit
+   * re-render the card with no write to the message.
+   */
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/zero/queries.ts]: remove query not needed, put logic in webhook to update this message

File: apps/dashboard/src/global.css [hunk 13]
```diff
@@ -5241,3 +5241,188 @@ img.inline-emoji,
...
 }
+
+/* ==========================================================================
+   Scheduled-call card (ScheduledCallPill)
+
+   Ported from the Oats design system's CallCard. The palette is carried as
+   scoped tokens rather than mapped onto our theme variables: the design system
+   uses its own neutral ramp (t1…t6) and a dedicated call-green family, and
+   approximating them with --foreground/--muted lost the intended look.
+   ========================================================================== */
```
Comment (Mayankbansal-juspay) [inline apps/dashboard/src/global.css]: too many global css changes please check once if needed or not.

File: apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.tsx [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+type PillCall = QueryResultType<typeof queries.scheduledCallPillCalls>[number];
+
+type PillState = 'SCHEDULED' | 'ACTIVE' | 'ENDED' | 'CANCELLED';
+
+/**
+ * Everything the card shows is derived from the live `calls` row rather than stored on
+ * the message, which is what lets a cancel (via the Zero mutator, which never reaches
+ * the backend) or a title/time edit re-render the card with no write to the message.
+ */
+const toPillState = (status: string | null | undefined): PillState => {
```
Comment (Mayankbansal-juspay) [inline apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.tsx]: pill state can be directly derived without this

File: apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.tsx [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+    case CallStatus.IN_PROGRESS:
+      return 'ACTIVE';
+    case CallStatus.ENDED:
+      return 'ENDED';
+    default:
+      return 'SCHEDULED';
+  }
+};
+
+const STATUS_LABEL: Record<PillState, string> = {
```
Comment (Mayankbansal-juspay) [inline apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.tsx]: put these in utils, or check for existing utils if those can be reused.

File: apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.tsx [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+    strokeLinejoin='round'
+    aria-hidden='true'
+  >
+    <rect x='3.5' y='4.5' width='13' height='12' rx='2' />
+    <path d='M3.5 8h13' />
+    <path d='M7 3.5v3M13 3.5v3' />
+  </svg>
+);
+
+const PhoneGlyph = (): React.JSX.Element => (
```
Comment (Mayankbansal-juspay) [inline apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.tsx]: these use through library which we already use, do not make these components

**Review by Mayankbansal-juspay — CHANGES_REQUESTED** [review]

File: apps/backend/src/database/repositories/callRepository.ts (hunk 11 above)
Comment (ishaanrawat-netizen) [reply apps/backend/src/database/repositories/callRepository.ts:98]: done

File: apps/backend/src/database/repositories/callRepository.ts [hunk 17]
```diff
@@ -1571,6 +1614,233 @@ export class CallRepository {
...
+        recurringSeriesId: true,
+        callType: true,
+      },
+    });
+
+    // A series is materialized 60 days ahead and replenished forever, so a pill per
+    // instance would drip cards into the channel indefinitely. Note-taker recordings
+    // are not meetings anyone joins. Enforced here rather than at the call sites: this
+    // is the single choke point, and it already has the row.
+    if (snapshot.isRecurring || snapshot.recurringSeriesId) return null;
```
Comment (XyneSpaces) [inline apps/backend/src/database/repositories/callRepository.ts:1668]: **[🟠 High] #83 — Pill gate lost the `callOrigin` check; the channel-move path can now mint pills for calendar-origin calls**

The original gate in this PR keyed on `callOrigin` — the doc comment added in `scheduleCallController.ts:38` still documents that intent ("The test is on `callOrigin`, not on having a channelId — `xyneManaged` calendar events carry a real one"). The rewrite dropped it: this choke point now only excludes recurring/series/HEADLESS calls.

The **create** path is safe — `scheduleCall` always passes origin CHANNEL/CONVERSATION. But `updateScheduledCall` → `moveScheduledCallPill` (line 1794) also funnels into `createScheduledCallPill`, and nothing on that path checks origin. A calendar-synced (EXTERNAL/`xyneManaged`) call — which carries a real channelId, per the comment above — that gets its channel edited would mint a pill it was never supposed to have.

**Rule:** removal of a guard without a code-path audit (#83 / R2).

**Fix:** add `callOrigin` to the `snapshot` select and return null unless it's in the Spaces-created set (CHANNEL/CONVERSATION) — the choke point then covers both create and move.

**Assumption:** `updateScheduledCall` is reachable for calendar-origin calls (participant/organizer authorization; no origin filter visible in the diff). If calendar-origin calls can never reach the move path this is moot — but the controller comment at :38 should still be corrected, since it documents a test that no longer exists.

File: apps/backend/src/database/repositories/callRepository.ts [hunk 18]
```diff
@@ -1571,6 +1614,233 @@ export class CallRepository {
...
+    tx: Prisma.TransactionClient,
+    params: {
+      messageId: string;
+      callExternalId: string;
+      movedToChannelId: string;
+      /** Snapshotted so the dead card still names the call it used to point at. */
+      callTitle?: string | null;
+    },
+  ): Promise<void> {
+    await tx.message.update({
```
Comment (XyneSpaces) [inline apps/backend/src/database/repositories/callRepository.ts:1773]: **[🟡 Medium] #59 — `retireScheduledCallPill` throws P2025 when the pill message was deleted, aborting the whole move transaction**

`tx.message.update` throws when `params.messageId` no longer exists. In `moveScheduledCallPill` the pill row is fetched at line 1816, but a null `pill` (message already deleted) falls through the `if (pill && …)` guard and still calls retire — so the entire `$transaction` rolls back and no fresh pill is minted in the destination channel. The controller catches and logs it, so the call edit itself survives, but the pill is silently lost.

Note `syncScheduledCallPillMessage` already guards this exact case (`if (!message) return;`, line 1747).

**Rule:** error handling for operations that can fail (#59 / R2).

**Fix:** make retire idempotent — `tx.message.updateMany({ where: { messageId }, data: … })` (no-throw on zero rows) — or in `moveScheduledCallPill` skip the retire call when `pill` is null, since the dead card is already gone.

File: apps/backend/src/database/repositories/callRepository.ts [hunk 19]
```diff
@@ -1571,6 +1614,233 @@ export class CallRepository {
...
+  /**
+   * Post the read-only "upcoming call" pill into the call's channel.
+   *
+   * A plain Prisma write, not a Zero mutator, so scheduling does not mark the channel
+   * unread for everyone — same as the existing call system message.
+   *
+   * Never writes `metadata.systemMessageId`: that belongs to `activateScheduledCall`,
+   * and stamping it would stop the live "X started a call" message from posting.
+   */
+  async createScheduledCallPill(
```
Comment (XyneSpaces) [inline apps/backend/src/database/repositories/callRepository.ts:1629]: **[🟠 High] #204 — No tests for the pill lifecycle (previously flagged, still open after the rewrite)**

The rewrite moved the whole lifecycle into the repository — `createScheduledCallPill` gating (recurring / series / HEADLESS / origin), `retireScheduledCallPill` semantics, `moveScheduledCallPill` transaction + thread guard, `syncScheduledCallPillMessage` retired/no-op branches, and the metadata merge that must preserve calendar fields and activation keys — and none of it has unit tests. Frontend side: `toPillState` / `formatCallWindow` / `PILL_STATUS_LABEL` and the ScheduledCallPill render states (scheduled / active / ended / cancelled / retired / no-snapshot) are likewise untested.

This was flagged at an earlier commit and the PR still contains no test files. The gate matrix in particular is exactly the kind of logic that regresses silently (see the origin-gate finding above).

**Rule:** #204 — add unit tests for new functionality and edge cases.

File: apps/backend/src/controllers/scheduleCallController.ts [hunk 20]
```diff
@@ -34,6 +35,11 @@ function defaultCallInvitation(timezone?: string | null): CallInvitationParams['
 }
 
 export class ScheduleCallController {
+  /**
```
Comment (XyneSpaces) [inline apps/backend/src/controllers/scheduleCallController.ts:38]: **[🟢 Low] #270 — Stale doc comment: documents a `callOrigin` test that no longer exists, above an unrelated method**

This block says "Whether a call gets a pill … The test is on `callOrigin`" — but (a) the pill gate now lives in `callRepository.createScheduledCallPill` and does **not** test `callOrigin`, and (b) it sits above `sendExternalInvitationInBackground`, which has nothing to do with pills.

**Rule:** leftover/stale comments (#270 / R8).

**Fix:** delete it — the gate in `callRepository.ts:1664` already carries an accurate comment — or move a corrected version there (especially if the origin check is re-added; see the High finding on the gate).

File: apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.css [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+  --cc-t6: #b3b9c0;
+  /* Call green — the design system's --call family, not --status-success. */
+  --cc-call: #57ab03;
+  --cc-call-hover: #4a9103;
+  --cc-call-text: #3f7a05;
+  --cc-call-bg: rgba(87, 171, 3, 0.07);
+  --cc-call-bd: rgba(87, 171, 3, 0.3);
+}
+
+[data-theme="midnight"] .xs-cc-scope {
```
Comment (XyneSpaces) [inline apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.css:22]: **[🟡 Medium] #182 — Only `midnight` gets a palette override; `summer_breeze` renders the raw Oats light palette (previously flagged, still open)**

The scoped-token approach with the ported design-system palette is reasonable and the header comment documents the deviation — but only `[data-theme="midnight"]` is overridden. In `summer_breeze` the card falls back to the pure-white/neutral Oats ramp, which may not sit right on that theme's surfaces.

**Rule:** theme-aware styling (#182).

**Fix:** either add a `[data-theme="summer_breeze"] .xs-cc-scope` block adjusting the few tokens that clash, or note in the PR that the default palette was verified against `summer_breeze` and passes.

File: apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.css [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+.xs-cc__btn:disabled {
+  opacity: 0.5;
+  cursor: not-allowed;
+}
+.xs-cc__btn--secondary {
+  background: var(--cc-bg);
+  color: var(--cc-t1);
+  border-color: var(--cc-bd);
+}
+.xs-cc__btn--secondary:hover:not(:disabled) {
```
Comment (XyneSpaces) [inline apps/dashboard/src/components/ui/MessageBubble/ScheduledCallPill.css:170]: **[🟢 Low] #155 — Buttons have `:hover` but no `:focus-visible` state (downgraded from the earlier flag)**

`.xs-cc__btn--secondary:hover:not(:disabled)` and `.xs-cc__btn--call:hover:not(:disabled)` restyle on hover, but nothing styles keyboard focus. Downgraded from the earlier Medium flag because the file never suppresses `outline` — the UA default ring still shows, so keyboard users do get feedback; this is visual-consistency polish, not an a11y hole.

**Rule:** #155 — pair `:focus` states with `:hover` states.

**Fix:** e.g. `.xs-cc__btn:focus-visible { outline: 2px solid var(--cc-call); outline-offset: 2px; }`.

**Review by XyneSpaces — COMMENTED** [review]
Re-review at `4470a2f9` after the repository-layer rewrite. Previously flagged items verified fixed: create-race (#105 — re-read inside the transaction), pill lifecycle moved into `callRepository`, `scheduledCallPillCalls` query removal, CSS scoping, utils extraction, lucide icons, dead disabled condition. Findings below — full summary in the top-level comment.

File: apps/backend/src/zero/queries.ts [hunk 23]
```diff
@@ -2927,6 +2927,7 @@ export const queries: AnyQueryRegistry = defineQueries({
     }
   ),
 
+
```
Comment (XyneSpaces) [inline apps/backend/src/zero/queries.ts:2930]: **[🟢 Low] #291 — Stray blank-line-only diffs in both zero query files**

This adds a second consecutive blank line before `userScheduledCalls` — and `packages/shared/src/zero/queries.ts:2382` picks up the identical change. Neither file has any functional change in this PR beyond this whitespace, so it reads as diff noise / accidental touch.

**Rule:** keep PRs focused / avoid unrelated changes (#291 / R9).

**Fix:** revert both hunks.

**Comment by XyneSpaces** [comment]
## Xyne Spaces Reviewer — PR #1972

**Verdict:** Changes requested
**Reviewed:** 13 files · 25 hunks @ `4470a2f9` · **Findings:** 2 high, 2 medium, 3 nits — 7 inline comments ([review](https://github.com/juspay/xyne-spaces/pull/1972#pullrequestreview-5266528593))

Re-review after the repository-layer rewrite. Previously flagged and now **verified fixed**: the create-race (#105 — re-read inside the transaction), pill lifecycle consolidated in `callRepository` (+ the fire-and-forget sync service), `scheduledCallPillCalls` query removal, CSS scoping, utils extraction, lucide icons, and the dead disabled condition. The rewrite is a clear improvement — what remains:

### Findings
| Severity | Rule | Location | Issue |
|----------|------|----------|-------|
| 🟠 High | #83 / R2 | `callRepository.ts:1668` | Pill gate lost the `callOrigin` check — the channel-move path can mint pills for calendar-origin (EXTERNAL/xyneManaged) calls |
| 🟠 High | #204 | `callRepository.ts:1629` | No unit tests for the pill lifecycle (gating, retire/move, sync, metadata merge, frontend state machine) — previously flagged, still open |
| 🟡 Medium | #59 / R2 | `callRepository.ts:1773` | `retireScheduledCallPill` throws P2025 if the pill message was deleted → whole move transaction aborts, no fresh pill |
| 🟡 Medium | #182 | `ScheduledCallPill.css:22` | Only `midnight` gets a palette override; `summer_breeze` renders the raw Oats light palette — previously flagged, still open |
| 🟢 Low | #270 | `scheduleCallController.ts:38` | Stale doc comment documents a `callOrigin` test that no longer exists, above an unrelated method |
| 🟢 Low | #155 | `ScheduledCallPill.css:170` | Buttons have `:hover` but no `:focus-visible` (UA ring preserved — consistency polish, downgraded) |
| 🟢 Low | #291 | `queries.ts:2930` (+ shared `:2382`) | Stray blank-line-only diffs in both zero query files |

### Recommended reviewer
**@mayankbansal-juspay** — deepest familiarity with this change: 10 detailed review comments today across 6 of the 13 changed files, and he drove the repository-layer redesign that this head implements.

| Candidate | Familiarity (this PR's files) | Default reviewer? | Pending review requests | Tie-breaker |
|-----------|-------------------------------|-------------------|-------------------------|-------------|
| @mayankbansal-juspay | 6 of 13 files, 10 review comments today | no | 330 of 495 open PRs (org automation auto-requests him broadly — not a human-load signal; he is actively reviewing today) | active on this PR today |
| @juspay/xyne-acl-guardians | CODEOWNERS of `apps/backend/src/zero/mutators.ts` (1 of 13 files) | yes (CODEOWNERS) | n/a — team, membership not enumerable via API | — |

Signal notes: the repo's CODEOWNERS covers only ACL paths via the `@juspay/xyne-acl-guardians` team — `mutators.ts` is changed here, so if branch protection enforces CODEOWNER review, a guardian still needs to sign off. Commits-by-path authorship could not be computed (no path-filtered commit API in this toolset), so familiarity is evidenced by review activity on this PR. The GitHub review-request API is not available to this agent, so @mayankbansal-juspay is tagged here rather than formally requested — please add him as a reviewer in the sidebar, or he can pick it up from this mention.

**Top blocker:** the origin-gate regression (`callRepository.ts:1668`) — a single `callOrigin` check at the choke point closes both the create and move paths.

**Review by Mayankbansal-juspay — DISMISSED** [review]

**Review by shubhamdhamija-juspay — DISMISSED** [review]

**Review by KushagarGoel1 — APPROVED** [review]
